SU2_CFD/include/numerics_simd/CNumericsSIMD.hpp

```diff
@@ -0,0 +1,711 @@
+  for (size_t iDim = 0; iDim < nDim; ++iDim) {
+    jac(0,iDim+1) = scale * normal(iDim);
+  }
+  jac(0,nDim+1) = 0.0;
+
+  for (size_t iDim = 0; iDim < nDim; ++iDim) {
+    jac(iDim+1,0) = scale * (normal(iDim)*phi - velocity[iDim]*projVel);
+    for (size_t jDim = 0; jDim < nDim; ++jDim) {
+      jac(iDim+1,jDim+1) = scale * (normal(jDim)*velocity[iDim] - gamma_m_1*normal(iDim)*velocity[jDim]);
+    }
+    jac(iDim+1,iDim+1) += scale * projVel;
+    jac(iDim+1,nDim+1) = scale * gamma_m_1 * normal(iDim);
+  }
+
+  jac(nDim+1,0) = scale * projVel * (phi-a1);
+  for (size_t iDim = 0; iDim < nDim; ++iDim) {
+    jac(nDim+1,iDim+1) = scale * (normal(iDim)*a1 - gamma_m_1*velocity[iDim]*projVel);
+  }
+  jac(nDim+1,nDim+1) = scale * gamma * projVel;
+
+  return jac;
+}
+
+template<size_t nVar>
+FORCEINLINE void updateLinearSystem(Int iEdge,
+                                    Int iPoint,
+                                    Int jPoint,
+                                    bool implicit,
+                                    UpdateType updateType,
+                                    const VectorDbl<nVar>& flux,
+                                    const MatrixDbl<nVar>& jac_i,
+                                    const MatrixDbl<nVar>& jac_j,
+                                    CSysVector<su2double>& vector,
+                                    CSysMatrix<su2mixedfloat>& matrix) {
+  if (updateType == UpdateType::COLORING) {
+    vector.AddBlock(iPoint, flux);
+    vector.SubtractBlock(jPoint, flux);
+    if(implicit) matrix.UpdateBlocks(iEdge, iPoint, jPoint, jac_i, jac_j);
+  }
+  else {
+    vector.SetBlock(iEdge, flux);
+    if (implicit) matrix.SetBlocks(iEdge, jac_i, jac_j);
+  }
+}
+
+/*!
+ * \class CRoeBase
+ * \brief Base class for Roe schemes, derived classes implement
+ *        the dissipation term in a "finalizeResidual" method.
+ */
+template<class Derived, size_t NDIM>
+class CRoeBase : public CNumericsSIMD {
+protected:
+  enum: size_t {nDim = NDIM};
+  enum: size_t {nVar = CCompressibleConservatives<nDim>::nVar};
+
+  const su2double kappa;
+  const su2double gamma;
+  const su2double entropyFix;
+  const bool dynamicGrid;
+
+  /*!
+   * \brief Constructor.
+   */
+  CRoeBase(const CConfig& config) :
+    kappa(config.GetRoe_Kappa()),
+    gamma(config.GetGamma()),
+    entropyFix(config.GetEntropyFix_Coeff()),
+    dynamicGrid(config.GetDynamic_Grid()) {
+
+  }
+
+public:
+  /*!
+   * \brief Implementation of the base Roe flux.
+   */
+  void ComputeFlux(unsigned long iEdge,
+                   const CConfig& config,
+                   const CGeometry& geometry,
+                   const CVariable& solution_,
+                   UpdateType updateType,
+                   CSysVector<su2double>& vector,
+                   CSysMatrix<su2mixedfloat>& matrix) const final {
+
+    const bool implicit = (config.GetKind_TimeIntScheme() == EULER_IMPLICIT);
+    const auto& solution = static_cast<const CEulerVariable&>(solution_);
```

Comment:
Another significant change from the current numerics is that (again for thread safety) we do not "set" anything into them, we pass them the whole variable structure (the solver `nodes`) and the numerics can read (and only read) any data it needs.
Similarly to what is done in the solvers, since we know the type of variable that goes with the numeric scheme we can `static_cast` to the derived type to avoid polymorphism.

Reply:
It will work since CEulerVariable is used in all of those in one way or another, i.e. CNSVariable is also a CEulerVariable with extra methods.